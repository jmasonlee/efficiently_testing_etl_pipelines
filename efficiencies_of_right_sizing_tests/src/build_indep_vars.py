"""
Program written by Jeff Levy (jlevy@urban.org) for the Urban Institute, last revised 8/24/2016.
Note that this is intended as a temporary work-around until pySpark improves its ML package.

Tested in pySpark 2.0.

"""
import pyspark


def build_indep_vars(df, independent_vars, categorical_vars=None, keep_intermediate=False, summarizer=True):
    check_input(categorical_vars, df, independent_vars)

    from pyspark.ml import Pipeline
    from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler

    if categorical_vars:
        string_indexer = [StringIndexer(inputCol=x,
                                        outputCol='{}_index'.format(x))
                          for x in categorical_vars]

        encoder        = [OneHotEncoder(dropLast=True,
                                        inputCol ='{}_index' .format(x),
                                        outputCol='{}_vector'.format(x))
                          for x in categorical_vars]

        independent_vars = ['{}_vector'.format(x) if x in categorical_vars else x for x in independent_vars]
    else:
        string_indexer, encoder = [], []

    assembler = VectorAssembler(inputCols=independent_vars,
                                outputCol='indep_vars')
    pipeline  = Pipeline(stages=string_indexer+encoder+[assembler])
    model = pipeline.fit(df)
    df = model.transform(df)

    if summarizer:
        param_crosswalk = {}

        i = 0
        for x in independent_vars:
            if '_vector' in x[-7:]:
                xrs = x.rstrip('_vector')
                dst = df[[xrs, '{}_index'.format(xrs)]].distinct().collect()

                for row in dst:
                    param_crosswalk[int(row['{}_index'.format(xrs)]+i)] = row[xrs]
                maxind = max(param_crosswalk.keys())
                del param_crosswalk[maxind] #for droplast
                i += len(dst)
            elif '_index' not in x[:-6]:
                param_crosswalk[i] = x
                i += 1

    if not keep_intermediate:
        fcols = [c for c in df.columns if '_index' not in c[-6:] and '_vector' not in c[-7:]]
        df = df[fcols]

    return df


def check_input(categorical_vars, df, independent_vars):
    assert (type(
        df) is pyspark.sql.dataframe.DataFrame), 'pypark_glm: A pySpark dataframe is required as the first argument.'
    assert (type(
        independent_vars) is list), 'pyspark_glm: List of independent variable column names must be the third argument.'
    for iv in independent_vars:
        assert (type(iv) is str), 'pyspark_glm: Independent variables must be column name strings.'
        assert (iv in df.columns), 'pyspark_glm: Independent variable name is not a dataframe column.'
    if categorical_vars:
        for cv in categorical_vars:
            assert (type(cv) is str), 'pyspark_glm: Categorical variables must be column name strings.'
            assert (cv in df.columns), 'pyspark_glm: Categorical variable name is not a dataframe column.'
            assert (cv in independent_vars), 'pyspark_glm: Categorical variables must be independent variables.'
