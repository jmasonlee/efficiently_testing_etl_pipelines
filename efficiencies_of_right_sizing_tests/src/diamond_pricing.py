from pyspark.sql import Window, DataFrame, Column
from pyspark.sql.functions import mean

from pyspark.sql.functions import when, col

def replace_null(orig: Column, average: Column):
    """replace_null

    Args:
        orig (Column): The original column
        average (double): the mean for that column

    Returns:
        Column: The original column with null values replaced by the average
    """
    return when(orig.isNull(), average).otherwise(orig)

def replace_null_prices_with_floating_averages(df: DataFrame) -> DataFrame:
    """replace_null_prices_with_floating_averages

    Args:
        df (DataFrame): A pyspark dataframe containing information about diamond prices. Expected format:
        
        +-----+---------+-----+-------+-----+-----+-----+----+----+----+
        |carat|      cut|color|clarity|depth|table|price|   x|   y|   z|
        +-----+---------+-----+-------+-----+-----+-----+----+----+----+
        | 0.23|    Ideal|    E|    SI2| 61.5| 55.0|  326|3.95|3.98|2.43|
        | 0.21|  Premium|    E|    SI1| 59.8| 61.0|  326|3.89|3.84|2.31|
        | 0.23|     Good|    E|    VS1| 56.9| 65.0| null|4.05|4.07|2.31|
        | 0.29|  Premium|    I|    VS2| 62.4| 58.0|  334| 4.2|4.23|2.63|
        | 0.31|     Good|    J|    SI2| 63.3| 58.0|  335|4.34|4.35|2.75|

    Returns:
        DataFrame: A dataframe where any null values in price have been replaced with 
        the average price for that day of all diamonds with the same cut, color, and clarity.
    """

    window = Window.partitionBy('cut', 'clarity').orderBy('price').rowsBetween(-3, 3)
    moving_avg = mean(df['price']).over(window)
    df = df.withColumn('moving_avg', moving_avg)

    df_new = df.withColumn('imputed', replace_null(col('price'), col('moving_avg')))

    return df_new
