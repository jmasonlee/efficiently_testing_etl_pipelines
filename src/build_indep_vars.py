"""
Program written by Jeff Levy (jlevy@urban.org) for the Urban Institute, last revised 8/24/2016.
Note that this is intended as a temporary work-around until pySpark improves its ML package.

Tested in pySpark 2.0.

"""
import pyspark


def build_indep_vars(df, independent_vars, categorical_vars=None, keep_intermediate=False, summarizer=True):
    check_input(categorical_vars, df, independent_vars)

    from pyspark.ml import Pipeline
    from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler
    idx = 'index'
    vec = 'vector'
    if categorical_vars:
        string_indexer = [StringIndexer(inputCol=x,
                                        outputCol=f"{x}_{idx}")
                          for x in categorical_vars]

        encoder        = [OneHotEncoder(dropLast=True,
                                        inputCol =f'{x}_{idx}',
                                        outputCol=f'{x}_{vec}')
                          for x in categorical_vars]

        independent_vars = ['{}_vector'.format(x) if x in categorical_vars else x for x in independent_vars]
    else:
        string_indexer, encoder = [], []

    assembler = VectorAssembler(inputCols=independent_vars,
                                outputCol='indep_vars')
    pipeline  = Pipeline(stages=string_indexer+encoder+[assembler])
    model = pipeline.fit(df)
    df = model.transform(df)

    if not keep_intermediate:
        fcols = [c for c in df.columns if f'_{idx}' not in c[-3:] and f'_{vec}' not in c[-7:]]
        df = df[fcols]

    return df


def check_input(categorical_vars, df, independent_vars):
    assert (type(
        df) is pyspark.sql.dataframe.DataFrame), 'pypark_glm: A pySpark dataframe is required as the first argument.'
    assert (type(
        independent_vars) is list), 'pyspark_glm: List of independent variable column names must be the third argument.'
    for iv in independent_vars:
        assert (type(iv) is str), 'pyspark_glm: Independent variables must be column name strings.'
        assert (iv in df.columns), 'pyspark_glm: Independent variable name is not a dataframe column.'
    if categorical_vars:
        for cv in categorical_vars:
            assert (type(cv) is str), 'pyspark_glm: Categorical variables must be column name strings.'
            assert (cv in df.columns), 'pyspark_glm: Categorical variable name is not a dataframe column.'
            assert (cv in independent_vars), 'pyspark_glm: Categorical variables must be independent variables.'
