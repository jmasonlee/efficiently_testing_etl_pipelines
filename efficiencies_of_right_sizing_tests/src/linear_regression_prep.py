from pyspark.sql import DataFrame
from pyspark.sql.functions import log

from efficiencies_of_right_sizing_tests.src.build_indep_vars import build_indep_vars


def transform(df: DataFrame) -> DataFrame:
    df = df[['carat', 'clarity', 'color', 'price']]

    df = df.withColumn('lprice', log('price'))
    df = replace_null_prices_with_floating_averages(df)
    df = build_indep_vars(df, ['carat', 'clarity', 'color'],
                                      categorical_vars=['clarity', 'color'],
                                      keep_intermediate=False,
                                      summarizer=True)
    return df
