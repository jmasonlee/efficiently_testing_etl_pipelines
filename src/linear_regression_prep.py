from pyspark.sql import DataFrame
from pyspark.sql.functions import log

from src.build_indep_vars import build_indep_vars
from src.diamond_pricing import replace_null_prices_with_floating_averages


def transform(df: DataFrame) -> DataFrame:

    df = df.withColumn('lprice', log('price'))
    df = replace_null_prices_with_floating_averages(df)
    df = df[['id', 'carat', 'clarity', 'color', 'price']]
    df = build_indep_vars(df, ['carat', 'clarity', 'color'],
                                      categorical_vars=['clarity', 'color'],
                                      keep_intermediate=False,
                                      summarizer=True)
    return df
