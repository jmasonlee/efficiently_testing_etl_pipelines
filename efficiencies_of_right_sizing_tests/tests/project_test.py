from src.project import replace_null_prices_with_floating_averages
from approvaltests import verify
from pyspark.sql import DataFrame

def test_will_replace_null_prices_with_floating_averages(spark) -> None:
    df = spark.createDataFrame([{"price": 327, "cut": "good", "clarity": "SI2"}])
    df = replace_null_prices_with_floating_averages(df)
    
    verify(get_string_representation_of_dataframe(df))

def get_string_representation_of_dataframe(df: DataFrame) -> str:
    str_df = df.toPandas().to_string(header=True, index=False)
    for name, dtype in df.dtypes:
        str_df = str_df.replace(name, f"{name}::{dtype}")
    return str_df