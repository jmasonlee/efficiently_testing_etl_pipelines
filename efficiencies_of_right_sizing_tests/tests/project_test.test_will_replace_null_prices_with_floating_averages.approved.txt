clarity::string  cut::string  price::bigint  moving_avg::double  imputed::double
    SI2 good    327       327.0    327.0
